src/commons/authentication/mod.rs

```diff
@@ -38,31 +37,106 @@ pub struct AuthenticationClassSpec {
 #[serde(rename_all = "camelCase")]
 #[allow(clippy::large_enum_variant)]
 pub enum AuthenticationClassProvider {
-    Ldap(LdapAuthenticationProvider),
-    Tls(TlsAuthenticationProvider),
-    Static(StaticAuthenticationProvider),
+    Static(static_::AuthenticationProvider),
+    Ldap(ldap::AuthenticationProvider),
+    Oidc(oidc::AuthenticationProvider),
+    Tls(tls::AuthenticationProvider),
 }
 
 impl AuthenticationClass {
     pub async fn resolve(
         client: &Client,
         authentication_class_name: &str,
-    ) -> Result<AuthenticationClass, Error> {
+    ) -> OperatorResult<AuthenticationClass> {
         client
             .get::<AuthenticationClass>(authentication_class_name, &()) // AuthenticationClass has ClusterScope
             .await
     }
 }
 
+/// Common [`ClientAuthenticationDetails`] which is specified at the client/
+/// product cluster level. It provides a name (key) to resolve a particular
+/// [`AuthenticationClass`]. Additionally, it provides authentication provider
+/// specific configuration (OIDC and LDAP for example).
+///
+/// If the product needs additional (product specific) authentication options,
+/// it is recommended to wrap this struct and use `#[serde(flatten)]` on the
+/// field.
+///
+/// Additionally, it might be the case that special fields are needed in the
+/// contained structs, such as [`oidc::ClientAuthenticationOptions`]. To be able
+/// to add custom fields in that structs without serde(flattening) multiple structs,
+/// they are generic, so you can add additional attributes if needed.
+///
+/// ### Example
+///
+/// ```
+/// # use schemars::JsonSchema;
+/// # use serde::{Deserialize, Serialize};
+/// use stackable_operator::commons::authentication::ClientAuthenticationDetails;
+///
+/// #[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+/// #[serde(rename_all = "camelCase")]
+/// pub struct SupersetAuthenticationClass {
+///     pub user_registration_role: String,
+///     pub user_registration: bool,
+///
+///     #[serde(flatten)]
+///     pub common: ClientAuthenticationDetails,
+/// }
+/// ```
+#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+#[schemars(description = "")]
+pub struct ClientAuthenticationDetails<O = ()> {
+    /// A name/key which references an authentication class. To get the concrete
+    /// [`AuthenticationClass`], we must resolve it. This resolution can be
+    /// achieved by using [`ClientAuthenticationDetails::resolve_class`].
+    #[serde(rename = "authenticationClass")]
+    authentication_class_ref: String,
+
+    /// This field contains authentication provider specific configuration. It
+    /// is flattened into the final CRD.
+    ///
+    /// Use [`ClientAuthenticationDetails::oidc_or_error`] to get the value or report an error to the user.
+    oidc: Option<oidc::ClientAuthenticationOptions<O>>,
```

Comment:
What happened here? This should be an enum with all supported authentication options:

```rust
pub enum ClientAuthenticationOptions {
    Ldap(ldap::ClientAuthenticationOptions),
    Oidc(oidc::ClientAuthenticationOptions),
}
```

Reply:
This is a breaking change, but I would have a really hard time pushing this OIDC-centric code. This kind of code gets forgotten and might block us to make changes in the future.

The good thing: Operators are not forced to use this common struct. It should be opt-in until we have CRD versioning in place.

Reply:
I tried my best, but could not get the enum to work...
I documented it in ab7b08edf5cc8b5a722c05f15a3b2c07e0cc35e5, would it be fine for you to continue like this for now?